### PR TITLE
Extract lib wave track

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: 

A toolkit-neutral library for the WaveTrack class and supporting classes.

But it still depends on a start-up dependency injection of the implementation
of block files.  That is, SqliteSampleBlock and SqliteSampleBlockFactory are
not in this library.  Only their abstract base classes.

So this library has no dependency on sqlite.

Notes for QA:  should test some of the GUI related to track displays, because this separates display from core logic.

- Display and clicks on clip boundaries and cutlines
- Adjustment of vertical bounds and scale type in vertical spectrogram ruler
- Varying spectrogram settings of individual tracks (NOT the use of the Preferences dialog which affects defaults)
- Waveform view vertical bounds and scale type changes

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/WaveTrack.cpp
Comment (crsib) [inline src/WaveTrack.cpp:175]: Why C-cast?

File: src/WaveTrack.cpp
Comment (crsib) [inline src/WaveTrack.cpp:185]: Why C-cast?

File: src/prefs/WaveformSettings.h
Comment (crsib) [inline src/prefs/WaveformSettings.h:93]: Something feels really wrong with this class. Judging by comment - it allows changing application global state. Following the real post effects is really hard here too.

File: src/prefs/WaveformSettings.h
Comment (crsib) [inline src/prefs/WaveformSettings.h:112]: Why setters are const?

File: src/prefs/WaveformSettings.h
Comment (crsib) [inline src/prefs/WaveformSettings.h:117]: Why the fields are public?

File: src/prefs/SpectrogramSettings.h
Comment (crsib) [inline src/prefs/SpectrogramSettings.h:203]: All the comments from the previous class apply here. 

BTW, why cache? What does it cache exactly?

File: src/WaveTrackLocation.cpp
Comment (crsib) [inline src/WaveTrackLocation.cpp:16]: Please use C++ header: cmath

**Review by crsib — CHANGES_REQUESTED** [review]
There are few places in this PR that need clarification

File: src/WaveTrack.cpp
Comment (Paul-Licameli) [reply src/WaveTrack.cpp:175]: Invoke assignment only if the ClientData::Site base class to clone the attachments.  That conserves the effect of the deleted lines.

File: src/WaveTrack.cpp
Comment (Paul-Licameli) [reply src/WaveTrack.cpp:185]: Ditto

File: src/WaveTrack.cpp
Comment (crsib) [reply src/WaveTrack.cpp:175]: I do not understand what exactly do you mean and how it justifies C-cast.

Comment (Paul-Licameli) [reply src/WaveTrack.cpp:175]: More explicitly `this->Caches::operator=(orig)` with no cast but I don’t find that prettier.

This is an intentional slicing assignment.

File: src/prefs/WaveformSettings.h
Comment (Paul-Licameli) [reply src/prefs/WaveformSettings.h:93]: Which comment do you mea?

This is extra state, but not of the application.  Each WaveTrack has its own independently changeable cache.

File: src/prefs/WaveformSettings.h
Comment (Paul-Licameli) [reply src/prefs/WaveformSettings.h:112]: See that the corresponding deleted member functions of WaveTrack were also like that.  The members were mutable.

Mutability makes some sense -- the graphics code wants to decorate a track, that it can examine but not change, with some extra memos only for its own use.

Instead of making those members of WaveTrackCache mutable, I could do this:

```
WaveformSettings &WaveformSettings::Get( const WaveTrack &track )
 {
+   return Get(const_cast<WaveTrack&>(track));
 }
```

-- that is, overload the static function accessing the cache, so it can take a const track but give back a non-const cache.

File: src/prefs/WaveformSettings.h
Comment (Paul-Licameli) [reply src/prefs/WaveformSettings.h:117]: No good reason here.  I can change it.

Though I do not agree that whenever there is a structure, you must encapsulate everything with get and set functions.  Some things are simply bags of data with no important consistency constraints, no invariants, to be established and preserved by member functions.  Such for instance is the case with many structure names ending in `...Options` that are just meant to shorten long lists of arguments.

File: src/prefs/SpectrogramSettings.h
Comment (Paul-Licameli) [reply src/prefs/SpectrogramSettings.h:203]: See what member variables are removed from WaveTrack.

They were all mutable and in that sense "caches."  Mutables aren't always nice but it seemed preferable to me before to make all those things mutables, rather than to require non-const tracks as arguments in drawing functions.  Logically, mere drawing should not be able to mutate the data model.  These extra things aren't really part of the data model.

Display and interaction code can take a const wave track that it is not allowed to mutate, but use ClientData to associate extra information with each track for its own needs.

File: src/WaveTrack.cpp
Comment (crsib) [reply src/WaveTrack.cpp:175]: I don't find this explanation convincing. This should be solved differently. Definitely not with a C-cast. 

I don't quite understand, why `static_cast` or some form of implicit conversion is not used here. 

I am not going to approve the code in the way it is now done.

File: src/prefs/WaveformSettings.h
Comment (crsib) [reply src/prefs/WaveformSettings.h:93]: Specifically this comment is really confusing.  `//! Get either the global default settings, or the track's own if previously created`

File: src/prefs/WaveformSettings.h
Comment (crsib) [reply src/prefs/WaveformSettings.h:117]: I have a pretty strong opinion, that mixing POD-like objects and getters/setters is not a best idea. Also in most of the places `m` prefix in Audacity means some kind of encapsulation. It is frustrating to read and understand the code when naming conventions are violated.

File: src/prefs/SpectrogramSettings.h
Comment (crsib) [reply src/prefs/SpectrogramSettings.h:203]: I don't see how making variable mutable makes it a `cache`.

File: src/WaveTrack.cpp
Comment (Paul-Licameli) [reply src/WaveTrack.cpp:175]: Okay then, will you be satisfied with:

`static_cast<Caches&>(*this) = orig;` ?

The reason C-cast didn't bother me is that this is an up-cast of the object to its base class, not the checked down-cast which is what `static_cast` is usually for with pointers and references.  In other words this also works with no casts:

```
Caches &caches = *this;
caches = orig;
```

In fact I even like that better.  No casts, just a reference binding.

Comment (Paul-Licameli) [reply src/WaveTrack.cpp:175]: As to what the line actually DOES, never mind the style matters -- there is no complaint?  It does no more nor less than the deleted lines it replaces.

File: src/prefs/WaveformSettings.h
Comment (Paul-Licameli) [reply src/prefs/WaveformSettings.h:93]: That comment may be extraneous, yes.  By mistaken analogy with the Spectogram settings code where that comment makes more sense.

File: src/prefs/WaveformSettings.h
Comment (Paul-Licameli) [reply src/prefs/WaveformSettings.h:117]: Arguably this is not a POD because there should be an ordering constraint on mins and maxes.

However I have my equally strong opinion that one should not be slavish about the `m` convention.  Some PODs really are PODs.  A structure with no invariant conditions it needs to protect is just a bunch of stuff and not properly a class.

File: src/prefs/SpectrogramSettings.h
Comment (Paul-Licameli) [reply src/prefs/SpectrogramSettings.h:203]: "memo" then?  Not in the sense of memo-izing an expensive computation, but it is some piece of associated attached data.  Or, if this helps, I guess in LISP world it would be something like a property list entry that you can stick onto a symbol.

"attachment"?

"decoration"?

other suggestions welcome

File: src/prefs/WaveformSettings.h
Comment (Paul-Licameli) [reply src/prefs/WaveformSettings.h:117]: The fact that something does NOT use the `m` conventions should be its own recognized convention too.  That, this structure is meant to be just a POD.

File: src/WaveTrack.cpp
Comment (crsib) [reply src/WaveTrack.cpp:175]: Implicit conversion, as with two lines you proposed, works just fine for me.

File: src/prefs/WaveformSettings.h
Comment (crsib) [reply src/prefs/WaveformSettings.h:117]: > Arguably this is not a POD because there should be an ordering constraint on mins and maxes.

Public access to members removes any constraints.

**Review by crsib — APPROVED** [review]

**Comment by Paul-Licameli** [comment]
Force pushed only to rebase onto recent master while waiting for QA approval.

**Comment by Ryan-Miller-Audacity** [comment]
QA verified on MacOs and Windows the following: 
* Display and clicks on clip boundaries and cutlines
* Adjustment of vertical bounds and scale type in vertical spectrogram ruler
* Varying spectrogram settings of individual tracks (NOT the use of the Preferences dialog which affects defaults)
* Waveform view vertical bounds and scale type changes

All appears to be working as expected when compared to previous versions.

**Pull request merged**
